src/Compilers/Core/Portable/RealParser.cs

```diff
@@ -0,0 +1,783 @@
+            // since the data are stored in quanta of 8 bits, and we really only need around 54  
+            // bits of mantissa for a double (and fewer for a float), we can just assemble data  
+            // from the eight high-order bytes and we will get between 59 and 64 bits, which is more  
+            // than enough.
+            int bottomElementIndex = Math.Max(0, topElementIndex - (64 / 8) + 1);
+            exponent = baseExponent + bottomElementIndex * 8;
+            mantissa = 0;
+            for (int i = (int)topElementIndex; i >= bottomElementIndex; i--)
+            {
+                mantissa <<= 8;
+                mantissa |= integerValueAsBytes[i];
+            }
+            for (int i = bottomElementIndex - 1; has_zero_tail && i >= 0; i--)
+            {
+                if (integerValueAsBytes[i] != 0) has_zero_tail = false;
+            }
+
+            return type.AssembleFloatingPointValue(mantissa, exponent, has_zero_tail, out result);
+        }
+
+        /// <summary>
+        /// Parse a sequence of digits into a BigInteger.
+        /// </summary>
+        /// <param name="data">The DecimalFloatingPointString containing the digits in its Mantissa</param>
+        /// <param name="integer_first_index">The index of the first digit to convert</param>
+        /// <param name="integer_last_index">The index just past the last digit to convert</param>
+        /// <returns>The BigInteger result</returns>
+        private static BigInteger AccumulateDecimalDigitsIntoBigInteger(DecimalFloatingPointString data, uint integer_first_index, uint integer_last_index)
+        {
+            if (integer_first_index == integer_last_index) return BigZero;
+            var valueString = data.Mantissa.Substring((int)integer_first_index, (int)(integer_last_index - integer_first_index));
+            return BigInteger.Parse(valueString);
+        }
+
+        /// <summary>
+        /// Return the number of significant bits set. 
+        /// </summary>
+        private static uint CountSignificantBits(ulong data)
+        {
+            uint result = 0;
+            while (data != 0)
+            {
+                data >>= 1;
+                result++;
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Return the number of significant bits set. 
+        /// </summary>
+        private static uint CountSignificantBits(byte data)
+        {
+            uint result = 0;
+            while (data != 0)
+            {
+                data >>= 1;
+                result++;
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Return the number of significant bits set. 
+        /// </summary>
+        private static uint CountSignificantBits(BigInteger data, out byte[] dataBytes)
+        {
+            if (data.IsZero)
+            {
+                dataBytes = new byte[1];
```

Comment:
This should occur rarely, if ever. It is an optimization over the following code as it stands.